### PR TITLE
Adds a marsh flooding coastal case

### PR DESCRIPTION
**Pull request by caozd999**
This PR adds two test cases that analyze the effects of salt marsh on coastal flooding. 
The cases implement marsh as a bottom drag in idealized land-coast-ocean transects. Each case includes two comparison experiments to show the effects of marsh width on coastal flooding.

### CODE REVIEW

**Comment by caozd999** [comment]
# testing
1. test case 52: -o ocean -c drying_slope -r marsh_flooding -t 100m
1.1 OB tidal amplitude check
tidalcomparison
1.2 Water elevation at different time
dryingslopecomparison

2. test case 53: -o ocean -c drying_slope -r marsh_flooding -t 250m
2.1 OB tidal amplitude check
tidalcomparison
2.2 Water elevation at different time
dryingslopecomparison

**Comment by caozd999** [comment]
> Probably don't need -- linked by COMPASS scripts.

File: src/core_ocean/mode_init/mpas_ocn_init_tidal_boundary.F [hunk 1]
```diff
@@ -261,6 +280,58 @@ subroutine ocn_init_setup_tidal_boundary(domain, iErr)!{{{
...
+            + (yCell(iCell) - yMin) / (yMax - yMin)  * &
+            (config_tidal_boundary_right_bottom_depth - config_tidal_boundary_left_bottom_depth)
+        end do
+      end if   !! if config_idealized_profile
+
+      if (config_tidal_boundary_right_bottom_depth < config_tidal_boundary_left_bottom_depth) then
+        call mpas_log_write('Right boundary must be deeper than left boundary!', MPAS_LOG_CRIT)
+      end if
+
+      ! Set refBottomDepth, bottomDepth, and maxLevelCell
```
Comment (pwolfram) [inline src/core_ocean/mode_init/mpas_ocn_init_tidal_boundary.F:330]: This is inconsistent with previous code: https://github.com/MPAS-Dev/MPAS-Model/pull/552/files#diff-6270014bc9ae9b8554f8af50127caf5fR263-R267 and seems to assume `config_tidal_boundary_left_bottom_depth = 0`.  The need to modify this is to accomodate other test cases of this type that may have `config_tidal_boundary_right_bottom_depth /=0`.

```
 ! Set refBottomDepth, bottomDepth, and maxLevelCell
      do k = 1, nVertLevels
        refBottomDepth(k) = config_tidal_boundary_left_bottom_depth + &
          (config_tidal_boundary_right_bottom_depth - config_tidal_boundary_left_bottom_depth)* interfaceLocations(k+1)
      end do
```

File: src/core_ocean/mode_init/mpas_ocn_init_tidal_boundary.F [hunk 2]
```diff
@@ -337,6 +408,7 @@ subroutine ocn_init_setup_tidal_boundary(domain, iErr)!{{{
       if (config_use_wetting_drying .and. config_tidal_start_dry) then
         do iCell = 1, nCellsSolve
           ssh(iCell) = -bottomDepth(iCell) + config_drying_min_cell_height*maxLevelCell(iCell)
+          ssh(iCell) = MAX(ssh(iCell),-config_tidal_forcing_monochromatic_baseline) !Zhendong Mar-06-2020
```
Comment (pwolfram) [inline src/core_ocean/mode_init/mpas_ocn_init_tidal_boundary.F:411]: @caozd999, can you please explain the purpose here?

File: src/core_ocean/mode_init/mpas_ocn_init_tidal_boundary.F [hunk 3]
```diff
@@ -363,14 +435,19 @@ subroutine ocn_init_setup_tidal_boundary(domain, iErr)!{{{
       ! Set tidal boundary mask
       do iCell = 1, nCellsSolve
         tidalInputMask(iCell) = 0.0_RKIND
-        if (yCell(iCell) > (25.0e3 - dcEdgeMinGlobal/2.0_RKIND)) then
+        if (yCell(iCell) > (yMax - dcEdgeMinGlobal/2.0_RKIND)) then ! Zhendong March-05-2020
```
Comment (pwolfram) [inline src/core_ocean/mode_init/mpas_ocn_init_tidal_boundary.F:438]: Nice change to generalize `yMax`.  No need for ` ! Zhendong March-05-2020` as it will be stored in git repo and can be found via a `git blame` on this line.

File: src/core_ocean/mode_init/mpas_ocn_init_tidal_boundary.F [hunk 4]
```diff
@@ -363,14 +435,19 @@ subroutine ocn_init_setup_tidal_boundary(domain, iErr)!{{{
...
           tidalInputMask(iCell) = 1.0_RKIND
         ! spread it over multiple cells
         !if (yCell(iCell) > (25.0e3 - 3*dcEdgeMinGlobal)) then
         !  tidalInputMask(iCell) = exp(-((yCell(iCell)-25.0e3)/dcEdgeMinGlobal)**2.0)
         end if
       end do
 
+      ! check that there is some tidalInputMask
```
Comment (pwolfram) [inline src/core_ocean/mode_init/mpas_ocn_init_tidal_boundary.F:446]: Nice!

File: src/core_ocean/shared/mpas_ocn_tidal_forcing.F [hunk 5]
```diff
@@ -263,7 +266,8 @@ subroutine ocn_tidal_forcing_build_array(domain, meshPool, forcingPool, diagnost
         ! compute the tidalHeight
         if (trim(config_tidal_forcing_model) == 'monochromatic') then
           tidalHeight = config_tidal_forcing_monochromatic_amp * &
-            SIN(2.0_RKIND*pii/config_tidal_forcing_monochromatic_period * daysSinceStartOfSim) - &
+            SIN(2.0_RKIND*pii/config_tidal_forcing_monochromatic_period * daysSinceStartOfSim - &
```
Comment (pwolfram) [inline src/core_ocean/shared/mpas_ocn_tidal_forcing.F:269]: Nice!   And verified that default registry value for `config_tidal_forcing_monochromatic_phaseLag` is 0.

File: testing_and_setup/compass/general.config.ocean [hunk 6]
```diff
@@ -9,23 +9,23 @@
 # init namelists in the default_inputs directory after a successful build of
 # the ocean model.
 [namelists]
```
Comment (pwolfram) [inline testing_and_setup/compass/general.config.ocean:11]: Please remove this file from your PR.  What I often do is have this in its own commit and use interactive rebase to keep it at the head, then when I push do something like this, where key thing is the `HEAD^` that removes this before you push:

`git push remote HEAD^:remote/branch/name`

File: testing_and_setup/compass/ocean/drying_slope/forward_template.xml [hunk 7]
```diff
@@ -26,7 +27,7 @@
 		<option name="config_use_variable_drag">.true.</option>
 		<option name="config_Rayleigh_friction">.true.</option>
 		<option name="config_Rayleigh_damping_depth_variable">.true.</option>
```
Comment (pwolfram) [inline testing_and_setup/compass/ocean/drying_slope/forward_template.xml:29]: @caozd999, we want this to be a ridiculous init value so that it breaks if not specified in a test case.  Is there a reason to set this to something particular?

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/comparison.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/env python
```
Comment (pwolfram) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/comparison.py:1]: I believe you can do this as a sym-link unless you've somehow modified this script from its original values.  If so, probably should generalize the other script into more modular code.

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/config_analysis.xml [hunk 9]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0"?>
+<config case="analysis">
+  <add_link source="../comparison.py" dest="comparison.py"/>
```
Comment (pwolfram) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/config_analysis.xml:3]: This can be a link to a more general location instead of it being hard-coded as done here.

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/config_driver.xml [hunk 10]
```diff
@@ -0,0 +1,17 @@
+<driver_script name="run_test.py">
+	<case name="init_step1">
+		<step executable="./run.py" quiet="true" pre_message=" * Running init_step1" post_message="     Complete"/>
+	</case>
+	<case name="forward1">
```
Comment (pwolfram) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/config_driver.xml:5]: Are each of these `forward1` and `forward2` cases needed?

Comment (pwolfram) [reply testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/config_driver.xml:5]: They appear to be the same below.

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/config_init1.xml [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+		<option name="config_idealized_profile_Sshore">0.001</option>
+		<option name="config_idealized_profile_Lmarsh">0.1</option>
+		<option name="config_idealized_profile_Smarsh">0.0</option>
+		<option name="config_idealized_profile_Cd">0.0025</option>
+		<option name="config_idealized_profile_Cd_marsh">0.25</option>
+		<option name="config_tidal_boundary_left_bottom_depth">0.0</option>
+		<option name="config_tidal_boundary_left_value">0.0</option>
+		<option name="config_tidal_boundary_right_value">2.5e4</option>
+		<option name="config_tidal_forcing_monochromatic_baseline">10.0</option>
+    <option name="config_alter_ICs_for_pbcs">.true.</option>
```
Comment (pwolfram) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/config_init1.xml:24]: MIssing tab character for indentation here...

Comment (pwolfram) [reply testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/config_init1.xml:24]: Note, have tabs show up in vim as special character for me because of things like this when editing xml.

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/config_init1.xml [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+  <add_link source="../../../meshes/100m/build_100m_mesh/graph.info" dest="graph.info"/>
+	<add_executable source="model" dest="ocean_model"/>
+
+	<namelist name="namelist.ocean" mode="init">
+		<template file="init_template.xml" path_base="script_configuration_dir"/>
+		<option name="config_use_variable_drag">.true.</option>
+		<option name="config_idealized_profile">.true.</option>
+		<option name="config_idealized_profile_Lcoast">0.3</option>
+		<option name="config_idealized_profile_Scoast">0.001</option>
+		<option name="config_idealized_profile_Lshore">0.6</option>
```
Comment (pwolfram) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/config_init1.xml:14]: Since this appears to be value that changed, recommend using templates so only this changed value shows up.  This was done for the WD cases for hurricanes, for example.

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/movie.py [hunk 13]
```diff
@@ -0,0 +1,213 @@
+#!/usr/bin/env python
```
Comment (pwolfram) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/movie.py:1]: Same comment -- can we make code more modular so we don't have copy/pasted script?  Not a big deal, but makes the repo nicer because if there is a bug, only have to fix it in a single place.  It also keeps code shorter too.

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/250m/config_analysis.xml [hunk 14]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0"?>
+<config case="analysis">
+  <add_link source="../comparison.py" dest="comparison.py"/>
```
Comment (pwolfram) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/250m/config_analysis.xml:3]: Likely should be a relative link.

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/250m/config_init1.xml [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+  <add_link source="../../../meshes/250m/build_250m_mesh/graph.info" dest="graph.info"/>
+	<add_executable source="model" dest="ocean_model"/>
+
+	<namelist name="namelist.ocean" mode="init">
+		<template file="init_template.xml" path_base="script_configuration_dir"/>
+		<option name="config_use_variable_drag">.true.</option>
+		<option name="config_idealized_profile">.true.</option>
+		<option name="config_idealized_profile_Lcoast">0.3</option>
+		<option name="config_idealized_profile_Scoast">0.001</option>
+		<option name="config_idealized_profile_Lshore">0.6</option>
```
Comment (pwolfram) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/250m/config_init1.xml:14]: Same comment as above-- should use a template for these `config_init*.xml` files.

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/config_forward1.xml [hunk 16]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0"?>
+<config case="forward1">
+	<add_link source="../init_step1/ocean.nc" dest="init.nc"/>
+	<add_link source="../init_step1/forcing.nc" dest="forcing.nc"/>
+	<add_link source="../init_step1/graph.info" dest="graph.info"/>
+
+	<namelist name="namelist.ocean" mode="forward">
+		<template file="forward_template.xml" path_base="script_configuration_dir"/>
+		<option name="config_Rayleigh_friction">.false.</option>
+		<option name="config_Rayleigh_damping_depth_variable">.false.</option>
```
Comment (pwolfram) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/config_forward1.xml:10]: Seems like template could be useful here.

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/config_forward2.xml [hunk 17]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0"?>
+<config case="forward2">
+	<add_link source="../init_step2/ocean.nc" dest="init.nc"/>
+	<add_link source="../init_step2/forcing.nc" dest="forcing.nc"/>
+	<add_link source="../init_step2/graph.info" dest="graph.info"/>
+
+	<namelist name="namelist.ocean" mode="forward">
```
Comment (pwolfram) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/config_forward2.xml:7]: Template would be good.

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/config_init2.xml [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+<config case="init_step2">
+
+  <add_link source="../../../meshes/100m/build_100m_mesh/100mmesh.nc" dest="mesh.nc"/>
+  <add_link source="../../../meshes/100m/build_100m_mesh/graph.info" dest="graph.info"/>
+	<add_executable source="model" dest="ocean_model"/>
+
+	<namelist name="namelist.ocean" mode="init">
+		<template file="init_template.xml" path_base="script_configuration_dir"/>
+		<option name="config_use_variable_drag">.true.</option>
+		<option name="config_idealized_profile">.true.</option>
```
Comment (pwolfram) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/100m/config_init2.xml:11]: Template?

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/250m/comparison.py [hunk 19]
```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/env python
```
Comment (pwolfram) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/250m/comparison.py:1]: Modularize?

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/250m/config_forward1.xml [hunk 20]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0"?>
+<config case="forward1">
+	<add_link source="../init_step1/ocean.nc" dest="init.nc"/>
+	<add_link source="../init_step1/forcing.nc" dest="forcing.nc"/>
+	<add_link source="../init_step1/graph.info" dest="graph.info"/>
+
+	<namelist name="namelist.ocean" mode="forward">
+		<template file="forward_template.xml" path_base="script_configuration_dir"/>
+		<option name="config_Rayleigh_friction">.false.</option>
+		<option name="config_Rayleigh_damping_depth_variable">.false.</option>
```
Comment (pwolfram) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/250m/config_forward1.xml:10]: Template?

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/250m/config_forward2.xml [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+<config case="forward2">
+	<add_link source="../init_step2/ocean.nc" dest="init.nc"/>
+	<add_link source="../init_step2/forcing.nc" dest="forcing.nc"/>
+	<add_link source="../init_step2/graph.info" dest="graph.info"/>
+
+	<namelist name="namelist.ocean" mode="forward">
+		<template file="forward_template.xml" path_base="script_configuration_dir"/>
+		<option name="config_Rayleigh_friction">.false.</option>
+		<option name="config_Rayleigh_damping_depth_variable">.false.</option>
+		<option name="config_use_implicit_bottom_drag">.false.</option>
```
Comment (pwolfram) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/250m/config_forward2.xml:11]: Template?

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/250m/config_init2.xml [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+  <add_link source="../../../meshes/250m/build_250m_mesh/250mmesh.nc" dest="mesh.nc"/>
+  <add_link source="../../../meshes/250m/build_250m_mesh/graph.info" dest="graph.info"/>
+	<add_executable source="model" dest="ocean_model"/>
+
+	<namelist name="namelist.ocean" mode="init">
+		<template file="init_template.xml" path_base="script_configuration_dir"/>
+		<option name="config_use_variable_drag">.true.</option>
+		<option name="config_idealized_profile">.true.</option>
+		<option name="config_idealized_profile_Lcoast">0.3</option>
+		<option name="config_idealized_profile_Scoast">0.001</option>
```
Comment (pwolfram) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/250m/config_init2.xml:13]: Template?

File: testing_and_setup/compass/ocean/drying_slope/marsh_flooding/250m/movie.py [hunk 23]
```diff
@@ -0,0 +1,213 @@
+#!/usr/bin/env python
```
Comment (pwolfram) [inline testing_and_setup/compass/ocean/drying_slope/marsh_flooding/250m/movie.py:1]: Modularize?

**Comment by pwolfram** [comment]
@caozd999, can you please do a rebase on `ocean/coastal`?

**Comment by pwolfram** [comment]
# Testing 250m

dryingslopecomparison

Result consistent with one shown above and will retest following PR edits.

File: src/core_ocean/mode_init/mpas_ocn_init_tidal_boundary.F (hunk 2 above)
Comment (caozd999) [reply src/core_ocean/mode_init/mpas_ocn_init_tidal_boundary.F:411]: In **shared/mpas_ocn_tidal_forcing.F**, the tidalHeight is computed as
>`tidalHeight = config_tidal_forcing_monochromatic_amp * &
             SIN(2.0_RKIND*pii/config_tidal_forcing_monochromatic_period *daysSinceStartOfSim - & pii*config_tidal_forcing_monochromatic_phaseLag/180.0_RKIND) - &
             config_tidal_forcing_monochromatic_baseline`

So adding the line 
>`ssh(iCell) = MAX(ssh(iCell),-config_tidal_forcing_monochromatic_baseline)` 

will ensure that tides propagate just on top of the initial ssh, instead of possibly having a distance above the initial ssh.

File: src/core_ocean/mode_init/mpas_ocn_init_tidal_boundary.F (hunk 3 above)
Comment (caozd999) [reply src/core_ocean/mode_init/mpas_ocn_init_tidal_boundary.F:438]: Will clean it up.

File: testing_and_setup/compass/ocean/drying_slope/forward_template.xml (hunk 7 above)
Comment (caozd999) [reply testing_and_setup/compass/ocean/drying_slope/forward_template.xml:29]: You are right, there is no reason to change this value.

**Comment by pwolfram** [comment]
Need to merge #595 before this case.

**Comment by sbrus89** [comment]
This has been replaced with #651.

**Comment by caozd999** [comment]
> This has been replaced with #651.

Thanks @sbrus89, I agree with it.
